### PR TITLE
[SPARK-12797] [SQL] Generated TungstenAggregate (without grouping keys)

### PR DESCRIPTION
**Pull request by davies**
As discussed in #10786, the generated TungstenAggregate does not support imperative functions.

For a query

```
sqlContext.range(10).filter("id > 1").groupBy().count()
```

The generated code will looks like:

```
/* 032 */     if (!initAgg0) {
/* 033 */       initAgg0 = true;
/* 034 */
/* 035 */       // initialize aggregation buffer
/* 037 */       long bufValue2 = 0L;
/* 038 */
/* 039 */
/* 040 */       // initialize Range
/* 041 */       if (!range_initRange5) {
/* 042 */         range_initRange5 = true;
       ...
/* 071 */       }
/* 072 */
/* 073 */       while (!range_overflow8 && range_number7 < range_partitionEnd6) {
/* 074 */         long range_value9 = range_number7;
/* 075 */         range_number7 += 1L;
/* 076 */         if (range_number7 < range_value9 ^ 1L < 0) {
/* 077 */           range_overflow8 = true;
/* 078 */         }
/* 079 */
/* 085 */         boolean primitive11 = false;
/* 086 */         primitive11 = range_value9 > 1L;
/* 087 */         if (!false && primitive11) {
/* 092 */           // do aggregate and update aggregation buffer
/* 099 */           long primitive17 = -1L;
/* 100 */           primitive17 = bufValue2 + 1L;
/* 101 */           bufValue2 = primitive17;
/* 105 */         }
/* 107 */       }
/* 109 */
/* 110 */       // output the result
/* 112 */       bufferHolder25.reset();
/* 114 */       rowWriter26.initialize(bufferHolder25, 1);
/* 118 */       rowWriter26.write(0, bufValue2);
/* 120 */       result24.pointTo(bufferHolder25.buffer, bufferHolder25.totalSize());
/* 121 */       currentRow = result24;
/* 122 */       return;
/* 124 */     }
/* 125 */
```

cc @nongli

### CODE REVIEW

**Comment by rxin** [comment]
What's the difference between this one and #10786?

**Comment by davies** [comment]
#10768 had more unrelated changes, is used for prototype, this is the one ready for review.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala [hunk 1]
```diff
@@ -35,4 +37,13 @@ class WholeStageCodegenSuite extends SparkPlanTest with SharedSQLContext {
       sortAnswers = false
     )
   }
+
+  test("Aggregate should be included in WholeStageCodegen") {
```
Comment (nongli) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/WholeStageCodegenSuite.scala]: can you add a case with multiple agg exprs?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala [hunk 2]
```diff
@@ -112,6 +113,103 @@ case class TungstenAggregate(
...
+      val initVars = if (e.nullable) {
+        s"""
+           | boolean $isNull = ${ev.isNull};
+           | ${ctx.javaType(e.dataType)} $value = ${ctx.defaultValue(e.dataType)};
+           | if (!${ev.isNull}) {
+           |   $value = ${ev.value};
+           | }
+         """.stripMargin
+      } else {
+        isNull = "false"
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala]: This is a bit intuitive: To pass "false" as the name of the variable when it is non-nullable. We should consider null or something like that (future patch).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala [hunk 3]
```diff
@@ -112,6 +113,103 @@ case class TungstenAggregate(
...
+         | }
+       """.stripMargin
+
+    (rdd, source)
+  }
+
+  override def doConsume(ctx: CodegenContext, child: SparkPlan, input: Seq[ExprCode]): String = {
+    // only have DeclarativeAggregate
+    val functions = aggregateExpressions.map(_.aggregateFunction.asInstanceOf[DeclarativeAggregate])
+    // the model could be only Partial or PartialMerge
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala]: model -> mode

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala [hunk 4]
```diff
@@ -112,6 +113,103 @@ case class TungstenAggregate(
...
+    val boundExpr = updateExpr.map(e => BindReferences.bindReference(e, inputAttr))
+    ctx.currentVars = bufVars ++ input
+    // TODO: support subexpression elimination
+    val codes = boundExpr.zipWithIndex.map { case (e, i) =>
+      val ev = e.gen(ctx)
+      if (e.nullable) {
+        s"""
+           | ${ev.code}
+           | ${bufVars(i).isNull} = ${ev.isNull};
+           | if (!${ev.isNull}) {
```
Comment (nongli) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregate.scala]: is this null check necessary? Probably faster to just assign always (maybe garbage) if that is safe.

**Comment by nongli** [comment]
Looking good, just some minor comments.

**Comment by davies** [comment]
The last commit had passed the tests, I'm going to merge this into master.

**Comment by rxin** [comment]
@davies it would be great if we can separate the generated code into two functions -- one that does the aggregation, and the other that does the output. This way, we can separate this into two "pipelines".

cc@nongli

**Comment by davies** [comment]
@rxin We can do that when this is a grouping key. For this case, it only output single row, usually it will be in the last few operators.

**Comment by rxin** [comment]
Why not just do it for both cases so it is more unified? I think the point is that we'd want the generated code to reflect more accurately the number of pipelines that are actually used.

**Comment by davies** [comment]
@rxin I see, it will be addressed by #10855

File: sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala [hunk 5]
```diff
@@ -71,7 +71,9 @@ class SQLMetricsSuite extends SparkFunSuite with SharedSQLContext {
       expectedNumOfJobs: Int,
       expectedMetrics: Map[Long, (String, Map[String, Any])]): Unit = {
     val previousExecutionIds = sqlContext.listener.executionIdToData.keySet
-    df.collect()
+    withSQLConf("spark.sql.codegen.wholeStage" -> "false") {
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: @davies, why this change?

Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala]: For this PR, the SQLMetrics are not supported in whole stage codegen.
